broker/query_plan.go

```diff
@@ -35,75 +46,138 @@ func (bpn *blockingPlanNodeImpl) Add(nodes ...common.BlockingPlanNode) {
 	nChildren := len(mn.children)
 	// checks before fan out
-	if common.Avg == mn.AggType {
+	if common.Avg == mn.aggType {
 		if nChildren != 2 {
 			err = utils.StackError(nil, "Avg MergeNode should have 2 children")
 			return
 		}
-		lhs, ok := mn.children[0].(*MergeNode)
+		lhs, ok := mn.children[0].(common.MergeNode)
 		if !ok {
 			err = utils.StackError(nil, "LHS of avg node must be sum node")
 			return
 		}
-		if common.Sum != lhs.AggType {
+		if common.Sum != lhs.AggType() {
 			err = utils.StackError(nil, "LHS of avg node must be sum node")
 			return
 		}
-		rhs, ok := mn.children[0].(*MergeNode)
+		rhs, ok := mn.children[1].(common.MergeNode)
 		if !ok {
 			err = utils.StackError(nil, "RHS of avg node must be count node")
 			return
 		}
-		if common.Count != rhs.AggType {
+		if common.Count != rhs.AggType() {
 			err = utils.StackError(nil, "RHS of avg node must be count node")
 			return
 		}
 	}
 
 	childrenResult := make([]queryCom.AQLQueryResult, nChildren)
-	errs := make([]error, nChildren)
+	errs := 0
 	wg := &sync.WaitGroup{}
 	for i, c := range mn.children {
 		wg.Add(1)
-		bpn, ok := c.(common.BlockingPlanNode)
-		if !ok {
-			err = utils.StackError(nil, "expecting BlockingPlanNode")
-			return
-		}
 		go func(i int, n common.BlockingPlanNode) {
 			defer wg.Done()
 			var res queryCom.AQLQueryResult
-			res, err = n.Run()
+			res, err = n.Run(ctx)
 			if err != nil {
 				// err means downstream retry failed
-				errs[i] = utils.StackError(err, "child node failed")
+				utils.GetLogger().With(
+					"error", err,
+				).Error("child node failed")
+				errs++
 				return
 			}
 			childrenResult[i] = res
-		}(i, bpn)
+		}(i, c)
 	}
 	wg.Wait()
 
+	if errs > 0 {
+		err = utils.StackError(nil, fmt.Sprintf("%d errors happend running merge node", errs))
+		return
+	}
+
 	result = childrenResult[0]
 	for i := 1; i < nChildren; i++ {
-		ctx := newResultMergeContext(mn.AggType)
-		result = ctx.run(result, childrenResult[i])
-		if ctx.err != nil {
-			err = ctx.err
+		mergeCtx := newResultMergeContext(mn.aggType)
+		result = mergeCtx.run(result, childrenResult[i])
+		if mergeCtx.err != nil {
+			err = mergeCtx.err
 			return
 		}
 	}
+	return
+}
+
+// ScanNode is a BlockingPlanNode that handles rpc calls to fetch data from datanode
+type ScanNode struct {
+	blockingPlanNodeImpl
+
+	query          query.AQLQuery
+	shardID        uint32
+	topo           topology.Topology
+	dataNodeClient dataCli.DataNodeQueryClient
+}
+
+func (sn *ScanNode) Run(ctx context.Context) (result queryCom.AQLQueryResult, err error) {
+	trial := 0
+	for trial < rpcRetries {
+		trial++
 
+		hosts, routeErr := sn.topo.Get().RouteShard(sn.shardID)
+		if routeErr != nil {
+			utils.GetLogger().With(
+				"error", routeErr,
+				"shard", sn.shardID,
+				"trial", trial).Error("failed to route shard")
+			err = utils.StackError(routeErr, "route shard failed")
+			continue
+		}
+		// pick random routable host
```

Comment:
avoid the one already failed?

Reply:
I thought about that, but seems unnessasary to remember the last host. the assumption here is that any unhealthy hosts will be removed from topology, then all hosts that are still in topology should be equal.

Reply:
fine for now, better we add the logic in the future, topology just use heartbeat, sometime error may due to other causes